### PR TITLE
Improve GetHashCode for IntPtr/UIntPtr on 64-bit.

### LINKED ISSUE #3506: Consider improving GetHashCode on IntPtr / UIntPtr for 64-bit values
The current implementation of `GetHashCode` on the `IntPtr` and `UIntPtr` types work by taking the value and converting it to an int, unchecked.

This works great on 32-bit values, but on 64-bit values, any values greater than the mask always result in the same hash code. For example:

```
var a = new IntPtr(0x123456FFFFFFFFL);
var b = new IntPtr(0x654321FFFFFFFFL);
Console.WriteLine(a.GetHashCode() == b.GetHashCode());
```

Which prints true. I would propose that for 64-bit values, it do something like:

```
var value = val.ToInt64();
var hash = unchecked((int)((value & 0xFFFFFFFFU) ^ ((value >> 32) & 0xFFFFFFFFU)));
```

That's an example, I'm sure people smarter than me have better ideas, or copy the implementation from `Int64`.

When `Size` 8 and keep the current implementation for when `Size` is 4.

Now, to preemptively answer the question *why*.

The obvious case is when these values are stored in a dictionary as keys. The next question might be, "Why are you using pointers as dictionary keys", to which I say, "I'm not". They are platform-dependent values, but they don't represent an address in memory or point to anything. They are just values. Why platform dependent integers? It was a strange choice of a library I am doing platform invoke with.

### PR DESCRIPTION
**Pull request by vcsjones**
The GetHashCode behavior on IntPtr and UIntPtr truncated the upper 32 bits.
This change uses the behavior of long and ulong, respectively, to generate
the hash code.

Fix #3506

### CODE REVIEW

**Comment by leppie** [comment]
:shipit:

**Comment by terrajobst** [comment]
Tagging a few folks:

@stephentoub @KrzysztofCwalina @jkotas

**Comment by jkotas** [comment]
The change looks good to me for CoreCLR, not sure whether we will take it for full .NET Framework. @vcsjones Could you please make this change under `FEATURE_CORECLR` ifdef?

**Comment by vcsjones** [comment]
@jkotas Sure thing. Is there a test I can use as an example for testing that it only impacts CoreCLR and not the Desktop CLR?

**Comment by jkotas** [comment]
The tests that we have in CoreCLR repo are CoreCLR-specific. We do not use them for full .NET Framework. You do not need to worry about making the test pass on full .NET Framework.

Some other comments:

- Our intend is to have all tests for framework surface in corefx - even though it may not be actually implemented in corefx. It would be nice to add this test to https://github.com/dotnet/corefx/blob/master/src/System.Runtime/tests/System/UIntPtr.cs. There is a couple of days lag between when change made in CoreCLR is picked up by corefx, so you need to wait a bit. Having similar test here is fine as well - we have similar duplication already.

- Assuming the fix goes through here, it should be done in CoreRT as well.

**Comment by stephentoub** [comment]
Also, I took a quick look at the corefx tests, and I don't believe any are explicitly testing for the exact current implementation, so this shouldn't break any existing tests (and if it happens to, we can fix those up quickly).

**Comment by vcsjones** [comment]
Thinking back on it, my second commit may not have the intended consequences. Is it fair to use `!FEATURE_CORECLR` to make something Core CLR specific, or is that not specific enough?

**Comment by jkotas** [comment]
`FEATURE_CORECLR` is fine. I just realized that you could make the 32-bit case better for CoreCLR as well - there is no good reason to strip the top bit. I would suggest to change it to
```
#if FEATURE_CORECLR
...improved version...
#else
...existing version...
#endif
```

**Comment by vcsjones** [comment]
Good point. Fixed.

**Comment by jkotas** [comment]
Also, could you please squash into a single commit?

File: src/mscorlib/src/System/IntPtr.cs [hunk 1]
```diff
@@ -109,7 +109,16 @@ public unsafe IntPtr(void* value)
...
-            return unchecked((int)((long)m_value));
+#if FEATURE_CORECLR
+    #if WIN32
+                return unchecked((int)((long)m_value));
+    #else
+                long l = (long)m_value;
+                return (unchecked((int)l) ^ (int)(l >> 32));
+    #endif
+#else
+                return unchecked((int)((long)m_value));
```
Comment (justinvp) [inline src/mscorlib/src/System/IntPtr.cs]: Nit: The indentation here is more than in `UIntPtr`. It'd be nice to be consistent.

Comment (vcsjones) [inline src/mscorlib/src/System/IntPtr.cs]: UIntPtr and IntPtr are fairly inconsistent with each other in regard to `ifdef`ed code. I tried to make the indentation match other places in IntPtr like ToInt32, but then the `ifdef` doesn't line up with that, either.

Is the correct spacing to have the `ifdef`s indented past the method body?

**Comment by vcsjones** [comment]
@jkotas No problem, I copied it verbatim from `Int64`'s GetHashCode, so I assumed the explicit conversion to `long` was there intentionally.

**Comment by vcsjones** [comment]
@jkotas The original implementation did the conversion to long. For the case where it isn't the Core CLR, should the long conversation be removed as well? I suspect it can be, but I'm never sure around corner-cases for compatibility.

**Comment by jkotas** [comment]
@vcsjones The !CoreCLR path should stay exactly same as before. As you have said, there can be corner cases that are hard to see...

**Comment by vcsjones** [comment]
OK, that's the latest. The change to UIntPtr respecting the top-most bit I don't think will break any tests in CoreFx after taking a glance.

**Comment by jkotas** [comment]
:+1:

**Comment by jkotas** [comment]
Would you mind submitting the same fix to CoreRT as well?

**Comment by vcsjones** [comment]
Yep. I'll tag you with the PR when I get it in.

**Comment by briansull** [comment]
@stephentoub @KrzysztofCwalina @jkotas

 I believe that the tests are not properly updated as PosTest2 is now failing for me:
ERROR!!!-002: expect IntPtr.GetHashCode() equals the address

if (ip.GetHashCode() != (int)mem)

**Comment by briansull** [comment]
Foor the UintPtr testthe failure is as follows:
Beginning scenario: PosTest3: UIntPtr with a value greater than Int32.MaxValue
ERROR!!!-005 TestId-P003: Actual hash code is not 2051597056 as expected: Actual(-95886592)

**Comment by vcsjones** [comment]
@briansull I'll examine.

**Comment by vcsjones** [comment]
The first comment @briansull mentioned is an issue and I will create a PR to fix it.

Regarding the UIntPtr one though, I'm unable to reproduce it on CoreCLR x64. Here were the steps I took to run it, these tests are a bit interesting to run, perhaps I am doing something wrong. Assume `C:\dev\coreclr` is the root of the repository.

```shell
build.cmd #Run build
msbuild /p:CLRTestPriorityToBuild=2 /t:rebuild UIntPtrGetHashCode.csproj #build project file
cd C:\dev\coreclr\bin\tests\Windows_NT.x64.Debug\CoreMangLib\cti\system\uintptr\UIntPtrGetHashCode
C:\dev\coreclr\bin\tests\Windows_NT.x64.Debug\Tests\Core_Root\CoreRun.exe UIntPtrGetHashCode.exe
```

The `CORE_ROOT` environment variable is set to "C:\dev\coreclr\bin\tests\Windows_NT.x64.Debug\Tests\Core_Root".

The failure is suspicious because `-95886592 & 0x7FFFFFFF` is 2051597056 which is part of the expected value.
